bliss/models/galaxy_net.py

```diff
@@ -110,50 +136,68 @@ def generate_latents(self, dataloader, n_batches):
+            recon_mean_main, pq_latent_main = self._main_forward(images, background)
+            loss_recon = self._get_likelihood_loss(images, recon_mean_main)
+            loss_prior = -pq_latent_main.sum()
+            loss = loss_recon + loss_prior
             self.log("train/loss_main", loss, prog_bar=True)
+            self.log("train/loss_main_recon", loss_recon)
         if optimizer_idx == 1:
             with torch.no_grad():
-                recon_mean_main = self._main_forward(images, background)
-            recon_mean_residual = self._residual_forward(images - recon_mean_main)
+                recon_mean_main, pq_latent_main = self._main_forward(images, background)
+            recon_mean_residual, pq_latent_residual = self._residual_forward(
+                images - recon_mean_main
+            )
             recon_mean_final = F.relu(recon_mean_main + recon_mean_residual)
-            loss = self._get_likelihood_loss(images, recon_mean_final)
+            loss_recon = self._get_likelihood_loss(images, recon_mean_final)
+            loss_prior = -(pq_latent_main.sum() + pq_latent_residual.sum())
+            loss = loss_recon + loss_prior
             self.log("train/loss", loss, prog_bar=True)
+            self.log("train/loss_recon", loss_recon, prog_bar=True)
         return loss
 
     def validation_step(self, batch, batch_idx):
         """Validation step (pytorch lightning)."""
         images, background = batch["images"], batch["background"]
-        recon_mean_main = self._main_forward(images, background)
-        loss_main = self._get_likelihood_loss(images, recon_mean_main)
+        recon_mean_main, p_latent_main = self._main_forward(images, background)
+        loss_recon_main = self._get_likelihood_loss(images, recon_mean_main)
+        loss_prior_main = -p_latent_main.sum()
+        loss_main = loss_recon_main + loss_prior_main
         self.log("val/loss_main", loss_main)
 
-        recon_mean_residual = self._residual_forward(images - recon_mean_main)
+        recon_mean_residual, p_latent_residual = self._residual_forward(images - recon_mean_main)
         recon_mean_final = F.relu(recon_mean_main + recon_mean_residual)
-        loss = self._get_likelihood_loss(images, recon_mean_final)
+        loss_recon = self._get_likelihood_loss(images, recon_mean_final)
+        loss_prior = loss_prior_main - p_latent_residual.sum()
+        loss = loss_recon + loss_prior
         self.log("val/loss", loss)
 
         # metrics
         residuals = (images - recon_mean_final) / torch.sqrt(recon_mean_final)
         residuals_main = (images - recon_mean_main) / torch.sqrt(recon_mean_main)
         recon_mean_residual = recon_mean_residual / torch.sqrt(recon_mean_main)
         self.log("val/max_residual", residuals.abs().max())
+
+        # Aggregate posterior
+        u_main = self.main_encoder(images - background)
+        u_residual = self.residual_encoder(images - recon_mean_main)
```

Comment:
what are the `u_main` and `u_residual` ?

Reply:
These are samples from the aggregate posterior (i.e. latent space samples from the real data). I renamed them to be more clear